packages/realm-react/src/useEmailPasswordAuth.tsx

```diff
@@ -0,0 +1,190 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { useApp, useAuthResult } from "./AppProvider";
+import { AuthOperationName, AuthResult } from "./types";
+import { Realm, User, Credentials } from "realm";
+import { useAuthOperation } from "./useAuthOperation";
+
+/**
+ * Hook providing operations and corresponding state for authenticating with a
+ * Realm app with Email/Password.  It also contains operations related to
+ * Email/Password authentication, such as resetting password and confirming a user.
+ *
+ * The {@link AuthResult} states returned from this hook are "global" for all
+ * components under a given AppProvider, as only one operation can be in progress
+ * at a given time (i.e. we will store the states on the context). This means that,
+ * for example, multiple components can use the `useAuth` hook to access
+ * `result.pending` to render a spinner when login is in progress, without
+ * needing to pass that state around or store it somewhere global in their app
+ * code.
+ */
+interface UseEmailPasswordAuth {
+  /**
+   * Convenience function to login a user with an email and password - users
+   * could also call `logIn(Realm.Credentials.emailPassword(email, password)).
+   *
+   * @returns A `Realm.User` instance for the logged in user.
+   */
+  logIn(args: { email: string; password: string }): Promise<Realm.User | void>;
+
+  /**
+   * Register a new user. By default this will login the newly registered user
+   * after they have been successfully registered, unless the
+   * `loginAfterRegister` property is `false`.
+   *
+   * @returns A `Realm.User` instance for the logged in user if
+   * `loginAfterRegister` is not `false`, `void` otherwise.
```

Comment:
Small improvement (which could be considered as a future enhancement): It might make sense to provide overloads for these two call patterns. That way a user wouldn't have to null check the returned promise value before accessing it if they don't explicitly provide `loginAfterRegister: false`.

Reply:
I'll have to give this some thought.  I mostly assume these hooks would be used in the `fallback` of `UserProvider`.  In this case, the actually returned result will never get used, since the component will be demounted by then.